Add license checking to the redact processor

joegallo opened the pull request:
Follow up to https://github.com/elastic/elasticsearch/pull/95426, and related to https://github.com/elastic/elasticsearch/issues/94882

Add a new `REDACT_PROCESSOR_FEATURE` which requires a `PLATINUM` license. Adds an optional `skip_if_unlicensed` configuration to the `redact` processor which defaults to ~`true`~ `false` (see #95928).

- If the current license level is sufficient for the processor, then constructing a processor and executing it work as before.
- Otherwise, the behavior depends on the value of `skip_if_unlicensed`:
  - If `skip_if_unlicensed` is `true` (the default), then the processor silently and 'successfully' does nothing.
  - If `skip_if_unlicensed` has been set to `false`, then the processor throws an exception.

The license checking logic runs at both runtime (for every document ingested) and at processor creation time (for every `redact` processor created). In a quick 'n dirty benchmark on my machine, it costs between 1-2 microseconds per check.

-----

Still to do:
- [x] Major:
  - [x] ~Switch the default value of `skip_if_unlicensed` to `false` per discussion in the comments below~ edit: That's actually a discussion item. I'm going to leave it as is and we change it on a different PR. double edit: Done, it was #95928.
  - [x] Need to handle the case of a node restart with an insufficient license -- the current code would throw an exception if `skip_if_unlicensed` is false, but I think we want to allow that work (and then fail at runtime on passed-in docs).
- [ ] Minor:
  - [x] Run down TODO comment in `XPackLicenseState`
  - [ ] Update the docs to reflect these changes

joegallo committed:
Add a license check

and the option to noop if the license check fails (which defaults to
true). The license check runs for every document processed by a redact
processor.

joegallo committed:
Add tests of the license checking

joegallo committed:
WIP add a todo

joegallo committed:
Add a license check on processor creation

And rewrite one of the RedactProcessor tests to get around the new
license check -- that test is specifically about testing the
post-construction runtime behavior.

joegallo committed:
Add tests of the on-creation license checking

dakrone commented on x-pack/plugin/redact/src/main/java/org/elasticsearch/xpack/redact/RedactProcessor.java:
```diff
@@ -341,6 +374,7 @@ public RedactProcessor create(
             String matchField = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, "field");
             List<String> matchPatterns = ConfigurationUtils.readList(TYPE, processorTag, config, "patterns");
             boolean ignoreMissing = ConfigurationUtils.readBooleanProperty(TYPE, processorTag, config, "ignore_missing", true);
+            boolean skipIfUnlicensed = ConfigurationUtils.readBooleanProperty(TYPE, processorTag, config, "skip_if_unlicensed", true);
```
I think I'd prefer to default this to `false`, what do you think?

joegallo replied:
Yeah, for sure, I was on the fence about what the right default value is. I'm happy to change it.

masseyke replied:
Yeah I'd think that for something that we advertise as being usable to redact PII, we wouldn't want it to silently no-op if someone temporarily mixed up a license.

davidkyle commented on x-pack/plugin/redact/src/main/java/org/elasticsearch/xpack/redact/RedactProcessor.java:
```diff
@@ -350,6 +384,12 @@ public RedactProcessor create(
             Map<String, String> customPatternBank = ConfigurationUtils.readOptionalMap(TYPE, processorTag, config, "pattern_definitions");
 
+            // on-creation license check, this is exercised at processor creation time
+            if (skipIfUnlicensed == false && REDACT_PROCESSOR_FEATURE.check(licenseState) == false) {
```
```suggestion
            if (REDACT_PROCESSOR_FEATURE.check(licenseState) == false) {
```

My understanding is that the `skip_if_unlicensed` is to prevent the redact processor failing the entire ingest pipeline once the licence expires. I'm less sure about allowing a user to optimistically create a processor with `skip_if_unlicensed` hoping it will become licensed later. 

It now strikes me that `skip_if_unlicensed` could be covered by the `ignore_failure` option that is common to all processors. To some, it may be confusing for the processor to fail a licence check if `ignore_failure: true`.

joegallo replied:
> I'm less sure about allowing a user to optimistically create a processor with skip_if_unlicensed hoping it will become licensed later.

That's pretty much the point of this work (https://github.com/elastic/elasticsearch/issues/94882), however -- if the `elastic/integrations` want to ship a `redact` processor by default, then we need to gracefully handle the license-not-sufficient case, since it could happen on every single document at scale and is not an unexpected case.

Another consideration is that the performance of `ignore_failure` isn't actually especially good, due to the implementation details of how it happens to be implemented (note, however, that that isn't set in stone, we could 'just' improve that).

In the same space -- `ignore_missing` (which many processors including redact already have) is also unnecessary, but we special case it for similar reasons. 

-----

All that said, though, I'm sympathetic. A more single-task-focused version of this PR could drop `skip_if_unlicensed` entirely, leaving that as an enhancement for another day. I'll see what kind of feedback I can get on going that way.

davidkyle replied:
>That's pretty much the point of this work (https://github.com/elastic/elasticsearch/issues/94882)

Thanks for the background, I see the need now

> ignore_missing ... is also unnecessary, but we special case it for similar reasons.

👍  it makes sense for `ignore_failure` and `skip_if_unlicensed` as distinct events

The InferenceProcessor could implement the same `skip_if_unlicensed` option by adding logic catch and unwrap the compliance exception and skip error reporting if `skip_if_unlicensed:true`. 

The trained model licensing error is thrown here: 
https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInternalInferModelAction.java#L125

davidkyle reviewed (approved):
LGTM

joegallo committed:
Add xpack licensing messages

joegallo committed:
Move the on-creation check to rest validation

Regardless of your current license, you need to be able to reboot. And
rebooting means that these processors will be re-built from the config
that's stored in the cluster state. So we can't have this validation
in the Factory's create method. So instead it's now magical 'extra
validation' that runs when you put a pipeline.

joegallo committed:
Merge branch 'main' into redact-processor-licensing

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Update docs/changelog/95477.yaml

masseyke commented on server/src/main/java/org/elasticsearch/ingest/Processor.java:
```diff
@@ -78,6 +78,18 @@ default boolean isAsync() {
+     * state B. We want to reject put requests on day 2 (at the rest/transport level), but still allow for restarting nodes in the
+     * cluster (so we can't throw exceptions from {@link Factory#create(Map, String, String, Map)}).
+     */
+    default void extraValidation() throws Exception {}
```
Maybe worth saying that implementations should throw an exception if there is a validation problem? Although I guess that's obvious since the return type is void.

joegallo replied:
Done, I added a little more onto the comment.

masseyke commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -81,6 +81,7 @@ public class XPackLicenseState {
                 "The CCR monitoring endpoint will be blocked",
                 "Existing follower indices will continue to replicate data" }
         );
+        messages.put(XPackField.REDACT_PROCESSOR, new String[] { "Executing a redact processor in an ingest pipeline will fail." });
```
" unless the `skipIfUnlicensed` flag is set to true, in which case the redact processor will be skipped"?

joegallo replied:
I'll do a bit of string review as a follow up -- the status quo in that file seems to favor pretty terse explanations, and I don't want to buck that, but I do also want to be truthful, etc.

masseyke reviewed (approved):
I can't think of a better approach, so LGTM.

joegallo commented on x-pack/plugin/redact/src/main/java/org/elasticsearch/xpack/redact/RedactProcessor.java:
```diff
@@ -75,10 +92,30 @@ public class RedactProcessor extends AbstractProcessor {
     @Override
     public IngestDocument execute(IngestDocument ingestDocument) {
+        // runtime license check, this runs for every document processed
+        if (REDACT_PROCESSOR_FEATURE.check(licenseState) == false) {
```
Yeah, it microbenchmarked at about 1 microseconds per doc on my machine. Just a fraction of a percent of the cost of actually running a redact processor on a document for real.

joegallo committed:
Elaborate a bit on exception throwing
